torch/distributed/_sharded_tensor/ops/linear.py

```diff
@@ -183,27 +197,42 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
         for idx, placement in enumerate(weight._sharding_spec.placements):
             split_size = input_split_sizes[placement.rank()]
             offset_start_idx = idx * sharded_dim_size_max
-            indices[placement.rank()] = list(range(offset_start_idx, offset_start_idx + split_size))
+            indices[placement.rank()] = list(
+                range(offset_start_idx, offset_start_idx + split_size)
+            )
         indices_flatten = list(idx for indice in indices for idx in indice)
 
-        input_t = input_t.index_select(0, torch.tensor(indices_flatten, device=input_t.device))
+        input_t = input_t.index_select(
+            0, torch.tensor(indices_flatten, device=input_t.device)
+        )
 
-    gathered_input = torch.empty(input_split_sizes[rank] * world_size, input_t_size[1], device=input_t.device)
+    output_tensor_list = [
+        torch.empty(
+            input_split_sizes[rank],
+            input_t_size[1],
+            device=input.device,
+            dtype=local_shard_t.dtype,
+            requires_grad=local_shard_t.requires_grad,
```

Comment:
looks like adding this does not help. Still get the None. Is there any way to print out the graph generated? cc: @pritamdamania87

Reply:
There are some details here: https://discuss.pytorch.org/t/print-autograd-graph/692 and https://stackoverflow.com/questions/52988876/how-can-i-visualize-what-happens-during-loss-backward in terms of how you can visualize/print the graph.

Reply:
Turns out that we need to call `local_shard.retain_grad()`.

Reply:
However, now the grad result does not match with the one from local. When print it out, the difference range from 0.2x to 2.2x.. So I doubt the correctness of the backward propagation. (But when running the original unit test -- test_all_to_all, it actually passes)

The local loss has the following graph:

> (None, 0)
> (<SumBackward1 object at 0x7f3d79327e80>, 0)
> (<AddmmBackward0 object at 0x7f3d79327ee0>, 0)
> (<AccumulateGrad object at 0x7f3d79327dc0>, 0)
> Tensor with grad found: torch.Size([11])
>  gradient: torch.Size([11])
> (None, 0)
> (<TBackward0 object at 0x7f3d79327df0>, 0)
> (<AccumulateGrad object at 0x7f3d79327eb0>, 0)
> Tensor with grad found: torch.Size([11, 16])
>  gradient: torch.Size([11, 16])

The one from sharding:

> (<MmBackward0 object at 0x7f35fb0a7880>, 0)
> (None, 0)
> (<CloneBackward0 object at 0x7f35fb0a7a30>, 0)
> (<TBackward0 object at 0x7f35fb0a77f0>, 0)
> (<CloneBackward0 object at 0x7f35fb0a7b20>, 0)
> (<SliceBackward0 object at 0x7f35fb0a7c70>, 0)
> (<AccumulateGrad object at 0x7f35fb0a7d60>, 0)
> Tensor with grad found: torch.Size([11, 16])
>  gradient: torch.Size([11, 16])
> (<AccumulateGrad object at 0x7f35fb0a7dc0>, 0)
> Tensor with grad found: torch.Size([11])
>  gradient: torch.Size([11])

This looks legit to me.. Maybe I need to look at it again tomorrow.